File: chapt2/chapt2_ws/src/demo_python_pkg/demo_python_pkg/learn_thread.py
import threading
import requests

class Download:
    def download(self, url, callback_word_count):
        print(f'线程:{threading.get_ident()} 开始下载:{url}')
        # 模拟下载过程
        response = requests.get(url)
        response.encoding = 'utf-8'
        callback_word_count(url,response.text)

    def start_download(self, url, callback_word_count):
        thread = threading.Thread(target=self.donload, args=(url, callback_word_count))
        thread.start()


def word_count(url, result):
    """
    普通函数,用于回调
    """
    print(f"{url}:{len(result)}->{result[:10]}") 


def main():
    download = Download()
    download.start_download('http://0.0.0.0:8000/novel1.txt', word_count)
    download.start_download('http://0.0.0.0:8000/novel2.txt', word_count)
    download.start_download('http://0.0.0.0:8000/novel3.txt', word_count)

File: chapt2/chapt2_ws/src/demo_cpp_pkg/src/learn_thread.cpp
#include <iostream>
#include <thread> //多线程
#include <chrono> //时间相关
#include <functional> //函数包装器
#include <cpp-httplib/httplib.h> //HTTP客户端

class Download
{
private:
    /* date */

public:
    void download(const std::string& host,const std::string& path, const 
    std::function<void(const std::string&, const std::string&)> &callback_word_count)
    {
        std::cout << "线程" << std::this_thread::get_id() << std::endl;
        httplib::Client client(host);
        auto response = client.Get(path);
        if(response && response->status == 200)
        {
            callback_word_count(path, response->body);
        }
    };

    void start_download(const std::string& host,const std::string& path, const 
    std::function<void(const std::string&, const std::string&)> &callback_word_count)
    {
        auto download_fun = std::bind(&Download::download, this,std::placeholders::_1,
        std::placeholders::_2, std::placeholders::_3);
        std::thread thread(download_fun, host, path, callback_word_count);
        thread.detach(); //分离线程，主线程结束后，子线程继续执行
    };
};

int main()
{
    auto d = Download();
    auto word_count = [](const std::string& path,const std::string& result) -> void
    {
        std::cout << "下载完成" << path << result.length() << result.substr(0,9) 
        << std::endl;
    };

    d.start_download("http://0.0.0.0:8000", "/novel1.txt", word_count);
    d.start_download("http://0.0.0.0:8000", "/novel2.txt", word_count);
    d.start_download("http://0.0.0.0:8000", "/novel3.txt", word_count);

    std::this_thread::sleep_for(std::chrono::milliseconds(1000*10));

    return 0;

}